Feature/application book cover supabase

AlexBud369 opened the pull request:
### Implemented Features

**Core Storage Changes**  
- Replaced Azure Blob Storage with Supabase Storage using S3-compatible API. **I decided to use supabase storage as a cloud storage service because I couldn't register on the azure portal or amazon (aws s3)**
- Added new `FileStorageService` implementing `IFileStorageService`  
- Configured Supabase settings in `appsettings.json` with placeholder values  

**Security Improvements**  
- Added `.gitignore` rules for sensitive files  
- Configured user secrets for Supabase keys  
- Implemented configuration validation  

**Localization Updates**  
- Updated shared resources for English/Russian  
- Standardized localization key access  
- Fixed `IStringLocalizer<SharedResources>` references  

**Type Safety Enhancements**  
- Introduced `BookSortBy` enum with serialization support  
- Refactored `UserRole` as proper enum  
- Updated domain constants with image extensions  

### Technical Details

**Storage Service**  
- Uses `AWSSDK.S3` NuGet package  
- Implements unique filename generation  
- Includes comprehensive logging

AlexBud369 committed:
chore: Update .gitignore to exclude Supabase secrets and sensitive files

AlexBud369 committed:
feat: add UserSecretsId to API.csproj for secure Supabase key storage

AlexBud369 committed:
feat: update appsettings.json with Supabase configuration and placeholders

AlexBud369 committed:
build: add AWSSDK.S3 package to Infrastructure for Supabase Storage integration

AlexBud369 committed:
refactor: Replace AzureFileStorageService with FileStorageService for Supabase Storage
- Replaced Azure Blob Storage (`BlobServiceClient`, `BlobContainerClient`) with Supabase Storage using S3-compatible API (`IAmazonS3`).
- Added comprehensive configuration validation for Supabase settings with localized error messages via `IStringLocalizer<SharedResources>`.
- Added unique file name generation with `public/` prefix to comply with Supabase Storage policies.

AlexBud369 committed:
feat: add PublicUrl to Supabase configuration

AlexBud369 committed:
feat: update SharedResources.en.resx and SharedResources.ru.resx with FileStorageService localization entries

AlexBud369 committed:
feat: rename SharedResource.cs to SharedResources.cs and add localization key constants

AlexBud369 committed:
feat: implement BookSortBy enum with serialization

AlexBud369 committed:
feat: Redesign UserRole enum with proper serialization

AlexBud369 committed:
chore: update DomainConstants with book cover extensions
- Added BookCover.AllowedImageExtensions (.jpg, .jpeg, .png)
- Removed deprecated Book.SortByTitle and Book.SortByCreatedAt
- Removed User.Roles as they're moved to UserRole enum"

AlexBud369 committed:
feat: add localization with new keys for handlers, services, and validators

AlexBud369 committed:
fix: Update DomainConstants.cs to reflect enum usage and pagination

budnikme commented on src/Infrastructure/Services/FileStorageService.cs:
```diff
@@ -0,0 +1,168 @@
+        ILogger<FileStorageService> logger,
+        IStringLocalizer<SharedResources> localizer)
+    {
+        Guard.AgainstNull(configuration, nameof(configuration), localizer.GetString(SharedResources.ConfigurationRequired));
```
These validations are not needed because .NET itself will throw an error if any of the services are not initialized properly

budnikme commented on src/Infrastructure/Services/FileStorageService.cs:
```diff
@@ -0,0 +1,168 @@
+        _logger = logger;
+        _localizer = localizer;
+
+        var (url, publicKey, secretKey, bucketName, publicUrl) = ValidateConfiguration(configuration, localizer);
```
Please use data annotations to validate configuration options https://www.roundthecode.com/dotnet-tutorials/validating-appsettings-faster-dotnet-8

budnikme commented on src/Infrastructure/Services/FileStorageService.cs:
```diff
@@ -0,0 +1,168 @@
+        ValidateConfiguration(
+        IConfiguration configuration, IStringLocalizer<SharedResources> localizer)
+    {
+        var url = configuration["Supabase:Url"];
```
Please inject configuration using The Options Pattern https://learn.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-9.0#the-options-pattern

budnikme commented on src/Infrastructure/Services/FileStorageService.cs:
```diff
@@ -0,0 +1,168 @@
+        var bucketName = configuration["Supabase:BucketName"];
+        var publicUrl = configuration["Supabase:PublicUrl"];
+
+        Guard.AgainstEmptyString(url, nameof(url), localizer.GetString(SharedResources.SupabaseUrlRequired));
```
Please use data annotations for these as well

budnikme commented on src/Infrastructure/Services/FileStorageService.cs:
```diff
@@ -0,0 +1,168 @@
+            BucketName = _bucketName,
+            Key = uniqueFileName,
+            InputStream = fileStream,
+            CannedACL = S3CannedACL.PublicRead
```
In most cases, for security reasons, ACL should always be private. If you really need it to be public in your case, add to your method an optional parameter with the default value of S3CannedACL.Private

AlexBud369 committed:
fix: Update Guard.cs to use SharedResources

AlexBud369 committed:
chore: add required NuGet packages to Infrastructure.csproj
- Added Amazon.S3 (0.34.0) for Supabase Storage integration.
- Added Microsoft.Extensions.Options (8.0.0) for Options Pattern in FileStorageService.
- Added Microsoft.Extensions.Localization (8.0.0) for IStringLocalizer<SharedResources>.
- Added Microsoft.Extensions.Logging.Abstractions (8.0.0) for ILogger<FileStorageService>.

AlexBud369 committed:
feat: update SharedResources and localization files with new keys

AlexBud369 committed:
feat: add localization keys

AlexBud369 committed:
feat: Add SupabaseOptions class for configuration

AlexBud369 committed:
fix: remove deprecated Amazon.S3 package from Infrastructure.csproj due to namespace conflicts

AlexBud369 committed:
fix: refactor FileStorageService to use SupabaseOptions and private uploads
- Replaced IConfiguration with IOptions<SupabaseOptions> for configuration validation
- Removed Guard.AgainstNull and Guard.AgainstEmptyString checks in favor of standard validation
- Changed file name prefix from public/ to private/ in GenerateUniqueFileName.
- Set CannedACL to S3CannedACL.Private in CreatePutObjectRequest for private uploads
- Removed ValidateConfiguration method as validation is handled by SupabaseOptions data annotations.
